### ISSUE TITLE
How to intercept all print task？

### ISSUE THREAD

**Issue by dreamtale90** [apple/cups#5845]
I want to intercept all the printing tasks of the operating system, judge the print files, release the files that are set to allow printing to continue printing, and stop printing the files that are set as not allowed to print. How can I implement it? The running environment is Linux

**Comment by jsmeix** [comment]
Hello,

On 2020-10-15 09:07, dreamtale90 wrote:
> I want to intercept all the printing tasks of the operating system,
> judge the print files, release the files that are set to allow
> printing to continue printing, and stop printing the files that are
> set as not allowed to print. How can I implement it? The running
> environment is Linux

I do not fully understand what you actually like to achieve in the end
but in general the basic idea for special print job handling
is to adapt and enhance the so called print job filtering.

For example you my have a look at
https://en.opensuse.org/SDB:Using_Your_Own_Filters_to_Print_with_CUPS
and
https://en.opensuse.org/SDB:Using_Your_Own_Backends_to_Print_with_CUPS

Both articles are a bit old so that this or that particular things
could be different with current CUPS versions but the general ideas
are still the same.


Kind Regards
Johannes Meixner
-- 
SUSE Software Solutions Germany GmbH
Maxfeldstr. 5 - 90409 Nuernberg - Germany
(HRB 36809, AG Nuernberg) GF: Felix Imendoerffer

**Comment by dreamtale90** [comment]
> Hello,
> On 2020-10-15 09:07, dreamtale90 wrote: I want to intercept all the printing tasks of the operating system, judge the print files, release the files that are set to allow printing to continue printing, and stop printing the files that are set as not allowed to print. How can I implement it? The running environment is Linux
> I do not fully understand what you actually like to achieve in the end but in general the basic idea for special print job handling is to adapt and enhance the so called print job filtering. For example you my have a look at https://en.opensuse.org/SDB:Using_Your_Own_Filters_to_Print_with_CUPS and https://en.opensuse.org/SDB:Using_Your_Own_Backends_to_Print_with_CUPS Both articles are a bit old so that this or that particular things could be different with current CUPS versions but the general ideas are still the same. Kind Regards Johannes Meixner -- SUSE Software Solutions Germany GmbH Maxfeldstr. 5 - 90409 Nuernberg - Germany (HRB 36809, AG Nuernberg) GF: Felix Imendoerffer

Thank you for your answer. Actually, I need to make a print control software. Files that are not allowed to be printed may be stored in a database (in which the relevant information of the file, file name, file path, hash value, etc.) are saved. After the software intercepts the print task, it can judge whether to continue printing operation by data query. If it is not matched in the database, it will continue to print, Stop printing if find it. That's about it.

**Comment by jschwender** [comment]
Sounds like Data Leakage Prevention, that is commercially available, you may look at something like itwatch.de. Usually, if there is a requirement for leakage prevention enforcement, just printer output does not cover it sufficiently. If it is a legal requirement you usually have do properly document violations and so on.

**Comment by debiantriage** [comment]
On Thu 15 Oct 2020 at 00:07:39 -0700, dreamtale90 wrote:

> I want to intercept all the printing tasks of the operating
> system,...

This is a job that Tea4CUPS can do very well:

  https://wiki.debian.org/CUPSTea4CUPS

> ...judge the print files, release the files that are set to allow
> printing to continue printing, and stop printing the files that
> are set as not allowed to print. How can I implement it? The
> running environment is Linux

You devise scripts for these procedures and tell Tea4CUPS to use
them.

Cheers,

Brian.

**Comment by dreamtale90** [comment]
Thank you for all the good advice. @jsmeix @jschwender @debiantriage

**Comment by Waxhaw** [comment]
> On Thu 15 Oct 2020 at 00:07:39 -0700, dreamtale90 wrote: I want to intercept all the printing tasks of the operating system,...
> This is a job that Tea4CUPS can do very well: https://wiki.debian.org/CUPSTea4CUPS
> ...judge the print files, release the files that are set to allow printing to continue printing, and stop printing the files that are set as not allowed to print. How can I implement it? The running environment is Linux
> You devise scripts for these procedures and tell Tea4CUPS to use them. Cheers, Brian.

Can anyone confirm that Tea4CUPS is still an active project?

I tried to get it to work on a Raspbian Pi using CUPS the latest which is:
CUPS 2.2.10
Tea4CUPS v3.14alpha

I kept getting this from the CUPS error logs:
D [04/Nov/2020:10:37:29 -0500] [Job 21] Tea4CUPS v3.14alpha_unofficial (PID 907) : I/O Error while waiting for lock /var/spool/cups/
D [04/Nov/2020:10:37:29 -0500] [Job 21] prehook_0 : cat $TEADATAFILE >/tmp/$TEAUSERNAME-$TEAJOBID.prn/tea4cups-..LCK
I [04/Nov/2020:10:37:30 -0500] Expiring subscriptions..

That seems very straight forward to how I was using a FILTER with just CUPS.

Anyone know where documents are on how to set this up correctly?  This project (Tea4CUPS) looks orphaned.

**Comment by jschwender** [comment]
Recent edits are some years old. Generally, Tea4CUPS uses raw queues, which are deprecated (#5271). That does affect future versions, but not the one you currently use (2.2.10). On the long run this looks like a dead end. But your error is about I/O, did you check access rights, mount options of that path?

**Comment by Waxhaw** [comment]
> Recent edits are some years old. Generally, Tea4CUPS uses raw queues, which are deprecated (#5271). That does affect future versions, but not the one you currently use (2.2.10). On the long run this looks like a dead end. But your error is about I/O, did you check access rights, mount options of that path?

I did but I will recheck this.  **Thank you for your quick response!**  I work for a digital photo printing company and they just want to provide AirPrint functionality but intercept the image to pass to their custom printing software built specifically for their printer.

So it looks like I was not correct on tea4cups.  Doing a search using the log file output for tea4cups showed me this code repot which has been updated in 8 months.

https://codeberg.org/dadosch/tea4cups

I'm going give that a shot and then post my results for others.
